kim/pipelines/base.py

```diff
@@ -36,21 +37,26 @@ def run(self, field, data, output):
 
         """
         pipe_data = data
-        for pipe in self.input_pipes:
-            pipe_data = pipe(field, pipe_data)
+        try:
+
+            for pipe in self.input_pipes:
+                pipe_data = pipe(field, pipe_data)
+
+            # validation pipes should not affect the data being piped
+            # they only raise exceptions if validation fails.
+            for pipe in self.validation_pipes:
+                pipe(field, pipe_data)
 
-        # validation pipes should not affect the data being piped
-        # they only raise exceptions of validation fails.
-        for pipe in self.validation_pipes:
-            pipe(field, pipe_data)
+            for pipe in self.process_pipes:
+                pipe_data = pipe(field, pipe_data)
 
-        for pipe in self.process_pipes:
-            pipe_data = pipe(field, pipe_data)
+            for pipe in self.output_pipes:
+                pipe(field, pipe_data, output)
```

Comment:
This is something that irks me greatly.

I think we need to find a way to treat all these things equally.  I think its going to be pretty confusing without some changes.  To answer your question directly the return value from output pipes would never be considered as `output` is passed along.